typing/includecore.ml

```diff
@@ -92,6 +93,9 @@ let primitive_descriptions pd1 pd2 =
 let value_descriptions ~loc env name
     (vd1 : Types.value_description)
     (vd2 : Types.value_description) =
+  (* modules are now fixed to be legacy *)
+  let mm1 = Mode.Value.legacy in
```

Comment:
I don't quite understand the semantics of these variables. The check here is a check between module *types*. How can we be talking about *modules*?

Reply:
A naive module inclusion check first checks that that the mode of the first module is less than the second. It then checks that the type of the first module is included by the second, which involves submodality check. Sub-modality, `f <= g`, is defined as "f x <= g x for all x". This is too strong - we already know the modes of modules. So instead we check that `f x <= g y` where `x` and `y` are the mode of first and second module.

Reply:
I don't think you want to treat:
`sig val foo : t end` and `sig val foo : t @@ global end` as equal module types just because all modules are currently global, but that is what this will do as written. You should pipe in `mm1` from the point at which you know you are comparing the module type of an actual module. When comparing module types from module type definitions then you should instead use top (or bottom? whichever one gives the right behaviour) rather than legacy.

Reply:
First off, I think there are now two different threads in here.

1. About the semantics of this function (my question).
2. How to behave in the current world where all modules are at the legacy mode (Leo's question).

I'll go back to 1. I had been thinking that the inclusion check was checking a subsumption relationship between two module types. But as I think about it more, I think that's wrong -- we never need to know one module type subsumes another, I think. (Importantly, the inclusion check treats module types themselves as invariant.) Instead, we need to know that some module `M : S1` can be used at type `S2`. So this gives a nice meaning for `mm1` here: it's the mode of `M` in that check. But what's `mm2` then?

Do you agree that we're always checking to see if some module `M` can be used at a type `S2`? And then can we get rid of `mm2` maybe? (Or perhaps it's the case that we use submoding to go from `mm1` to `mm2`... and one might be more permissive than the other regarding modality inclusion? Not sure here.)

Reply:
`mm1` is the mode of the module on LHS, `mm2` is the mode of the module on RHS.

We want to check that a module of type `S1` at mode `mm1`, can be used as a module of type `S2` at mode `mm2`. This involves:

1. First check that `mm1 <= mm2`. 
2. Then for each value, `f1 mm1 <= f2 mm2`, where `f1` and `f2` are the modalities on values.

1 is not needed right now because they are all legacy. 2 is what the code in question is about.

Reply:
OK. I'm satisfied now. I think the key thing for me is that this isn't a check about module types, despite the names of the functions that implement the check. It's a check to see whether a coercion on a particular module is safe. Thanks.

Reply:
Yeah, that's at least what I think. But maybe @lpw25 thinks differently.

Reply:
This function is used both for checking a relationship between two modules *and* for checking that two module types are equal (by checking the subsumption both ways). My point is that it is currently incorrect for that second use case. If you thread `mm1` and `mm2` in from the caller than you can get it to behave correctly for both uses:
- You can set them to the actual module modes in the first use case (which are currently both legacy)
- You can set them to top/bottom in the module type comparison case.

Reply:
Resolved since we had offline discussion.